File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultMetricsFactory.java
package com.krillsson.sysapi.core.metrics.defaultimpl;

import com.krillsson.sysapi.core.metrics.MetricsFactory;
import com.krillsson.sysapi.core.SpeedMeasurementManager;
import com.krillsson.sysapi.core.metrics.*;
import com.krillsson.sysapi.util.Utils;
import oshi.hardware.HardwareAbstractionLayer;
import oshi.software.os.OperatingSystem;

public class DefaultMetricsFactory implements MetricsFactory {

    private final HardwareAbstractionLayer hal;
    private final OperatingSystem operatingSystem;
    private final SpeedMeasurementManager speedMeasurementManager;
    private final Utils utils;

    private CpuMetrics cpuMetrics;
    private NetworkMetrics networkMetrics;
    private GpuMetrics gpuMetrics;
    private DiskMetrics diskMetrics;
    private ProcessesMetrics processesMetrics;
    private MotherboardMetrics motherboardMetrics;
    private MemoryMetrics memoryMetrics;

    public DefaultMetricsFactory(HardwareAbstractionLayer hal, OperatingSystem operatingSystem, SpeedMeasurementManager speedMeasurementManager, Utils utils) {
        this.hal = hal;
        this.operatingSystem = operatingSystem;
        this.speedMeasurementManager = speedMeasurementManager;
        this.utils = utils;
    }

    @Override
    public boolean prerequisitesFilled() {
        return true;
    }

    @Override
    public boolean initialize() {
        setCpuMetrics(new DefaultCpuMetrics(hal, operatingSystem, new DefaultCpuSensors(hal), utils));
        DefaultNetworkMetrics networkInfoProvider = new DefaultNetworkMetrics(hal, speedMeasurementManager);
        networkInfoProvider.register();
        setNetworkMetrics(networkInfoProvider);
        setGpuMetrics(new DefaultGpuMetrics(hal));
        DefaultDiskProvider diskInfoProvider = new DefaultDiskProvider(operatingSystem, hal, speedMeasurementManager);
        diskInfoProvider.register();
        setDiskMetrics(diskInfoProvider);
        setProcessesMetrics(new DefaultProcessesMetrics(operatingSystem, hal));
        setMotherboardMetrics(new DefaultMotherboardMetrics(hal));
        setMemoryMetrics(new DefaultMemoryMetrics(hal));
        return true;
    }

    @Override
    public CpuMetrics cpuInfoProvider() {
        return cpuMetrics;
    }

    @Override
    public NetworkMetrics networkInfoProvider() {
        return networkMetrics;
    }

    @Override
    public DiskMetrics diskInfoProvider() {
        return diskMetrics;
    }

    @Override
    public MemoryMetrics memoryInfoProvider() {
        return memoryMetrics;
    }

    @Override
    public ProcessesMetrics processesInfoProvider() {
        return processesMetrics;
    }

    @Override
    public GpuMetrics gpuInfoProvider() {
        return gpuMetrics;
    }

    @Override
    public MotherboardMetrics motherboardInfoProvider() {
        return motherboardMetrics;
    }

    protected void setCpuMetrics(CpuMetrics cpuMetrics) {
        this.cpuMetrics = cpuMetrics;
    }

    protected void setNetworkMetrics(NetworkMetrics networkMetrics) {
        this.networkMetrics = networkMetrics;
    }

    protected void setGpuMetrics(GpuMetrics gpuMetrics) {
        this.gpuMetrics = gpuMetrics;
    }

    protected void setDiskMetrics(DiskMetrics diskMetrics) {
        this.diskMetrics = diskMetrics;
    }

    protected void setProcessesMetrics(ProcessesMetrics processesMetrics) {
        this.processesMetrics = processesMetrics;
    }

    protected void setMotherboardMetrics(MotherboardMetrics motherboardMetrics) {
        this.motherboardMetrics = motherboardMetrics;
    }

    protected void setMemoryMetrics(MemoryMetrics memoryMetrics) {
        this.memoryMetrics = memoryMetrics;
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultMetricsFactory.kt
package com.krillsson.sysapi.core.metrics.defaultimpl

import com.krillsson.sysapi.config.SysAPIConfiguration
import com.krillsson.sysapi.core.connectivity.ConnectivityCheckManager
import com.krillsson.sysapi.periodictasks.TaskManager
import com.krillsson.sysapi.util.asOperatingSystem
import com.krillsson.sysapi.util.asPlatform
import oshi.PlatformEnum
import oshi.hardware.HardwareAbstractionLayer
import oshi.software.os.OperatingSystem

object DefaultMetricsFactory {
    fun create(
        config: SysAPIConfiguration,
        os: OperatingSystem,
        hal: HardwareAbstractionLayer,
        platformEnum: PlatformEnum,
        periodicTaskManager: TaskManager,
        diskReadWriteRateMeasurementManager: DiskReadWriteRateMeasurementManager,
        networkUploadDownloadRateMeasurementManager: NetworkUploadDownloadRateMeasurementManager,
        connectivityCheckManager: ConnectivityCheckManager
    ): DefaultMetrics {
        val defaultCpuLoadMetrics = DefaultCpuLoadMetrics(hal.processor, periodicTaskManager)
        val defaultCpuSensors = DefaultCpuSensors(hal)
        val cpuMetrics = DefaultCpuMetrics(hal, os, defaultCpuSensors, defaultCpuLoadMetrics)
        val diskMetrics = DefaultDiskMetrics(hal, diskReadWriteRateMeasurementManager)
        val fileSystemMetrics = DefaultFileSystemMetrics(os)
        val networkMetrics = DefaultNetworkMetrics(
            hal,
            networkUploadDownloadRateMeasurementManager,
            connectivityCheckManager
        )
        val gpuMetrics = DefaultGpuMetrics(hal)
        val processesMetrics = DefaultProcessesMetrics(config.processes, os, hal, periodicTaskManager)
        val motherboardMetrics = DefaultMotherboardMetrics(hal)
        val memoryMetrics = DefaultMemoryMetrics(hal, os)
        val systemMetrics = DefaultSystemMetrics(
            cpuMetrics,
            networkMetrics,
            diskMetrics,
            fileSystemMetrics,
            memoryMetrics,
            processesMetrics,
            motherboardMetrics,
            gpuMetrics,
            os.asOperatingSystem(),
            platformEnum.asPlatform()
        )
        return DefaultMetrics(
            cpuMetrics,
            networkMetrics,
            gpuMetrics,
            diskMetrics,
            fileSystemMetrics,
            processesMetrics,
            motherboardMetrics,
            memoryMetrics,
            systemMetrics
        )
    }
}